### PR TITLE
Implement Memcmp SIMD for arm64 NEON and SVE

### PR DESCRIPTION
**Pull request by AGSaidi**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://suricata.readthedocs.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
A request has been sent to add me to our corporate contribution agreemen
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)
NA

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- lengthen memcmp test input length to exercise simd instructions
- add timing assembly for arm64
- implement Memcmp SIMD for arm64 NEON and SVE

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the pull request number.

```
SV_REPO=
SV_BRANCH=
SU_REPO=
SU_BRANCH=
LIBHTP_REPO=
LIBHTP_BRANCH=
```

### CODE REVIEW

**Comment by victorjulien** [comment]
Could you also update the `SIMD support` line in `--build-info`?

E.g.
```
./src/suricata --build-info|grep SSE
SIMD support: SSE_4_2 SSE_4_1 SSE_3
```

Did you test this with `suricata-verify`? We have no arm64 running in our public CI, and I think my personal arm64 devices have no NEON/SVE, although I'm not quite sure. My most modern runner (rock 5b) has A76 coders. Not sure if that is compatible.

Formatting will need to be fixed.

**Comment by victorjulien** [comment]
Additionally, I would love to see to benchmarks on what the effect of using the SIMD is here.

**Comment by AGSaidi** [comment]
victorjulien@ anything else you’d like to see?

File: src/util-memcmp.c [hunk 1]
```diff
@@ -35,8 +35,8 @@
 
 static int MemcmpTest01 (void)
 {
-    uint8_t a[] = "abcd";
-    uint8_t b[] = "abcd";
+    uint8_t a[] = "abcd            ";
+    uint8_t b[] = "abcd            ";
```
Comment (catenacyber) [inline src/util-memcmp.c:39]: Why do we need all tests to have lengths of 16 ?

Comment (AGSaidi) [reply src/util-memcmp.c:39]: Values less than the SIMD width fall back to the simple scaler code and don’t test the more complex parts of the implementation.

Comment (victorjulien) [reply src/util-memcmp.c:39]: Agree, we shouldn't update existing tests for this, but instead add a new set of tests.

**Review by catenacyber — CHANGES_REQUESTED** [review]
see comments inline about tests

**Comment by AGSaidi** [comment]
thanks for the ping. I'll improve them after the holidays and resubmit.

**Comment by catenacyber** [comment]
Could you please resubmit as a new rebased PR ? This is Suricata's Github workflow cf https://docs.suricata.io/en/latest/devguide/contributing/code-submission-process.html#pull-requests

**Comment by catenacyber** [comment]
Closing as stale, please feel free to reopen a new rebased PR with the suggestions.

File: src/util-memcmp.h [hunk 2]
```diff
@@ -282,6 +282,94 @@ static inline int SCMemcmpLowercase(const void *s1, const void *s2, size_t len)
...
+
+#endif
+
+#define SCMEMCMP_BYTES 16
+
+#define UPPER_LOW   0x41 /* "A" */
+#define UPPER_HIGH  0x5A /* "Z" */
+#define UPPER_DELTA 0x20
+
+static inline int SCMemcmpLowercase(const void *s1, const void *s2, size_t len)
```
Comment (victorjulien) [inline src/util-memcmp.h:322]: it looks like this only uses NEON instructions, so it wouldn't see SVE?

Comment (AGSaidi) [reply src/util-memcmp.h:322]: the other branch of the `#if` uses SVE, this is pure NEON.

**Pull request closed without merging**